src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ForEachStatement.cs

```diff
@@ -621,6 +627,93 @@ private BoundStatement RewriteForEachStatementAsFor(BoundForEachStatement node,
+                                                                            }
+                                                                        }
+
+                                                                        collectionTempRefKind = enumeratorInfo.InlineArraySpanType == WellKnownType.System_Span_T ? RefKind.Ref : RefKindExtensions.StrictIn;
+                                                                        return collectionVarInitializationPreamble;
+                                                                    },
+                                                getItem: static (LocalRewriter rewriter, BoundForEachStatement node, BoundLocal boundArrayVar, BoundLocal boundPositionVar, object? _) =>
+                                                                {
+                                                                    MethodSymbol elementRef;
+                                                                    Debug.Assert(rewriter._factory.ModuleBuilderOpt is { });
+                                                                    Debug.Assert(rewriter._diagnostics.DiagnosticBag is { });
+
+                                                                    var enumeratorInfo = node.EnumeratorInfoOpt;
+                                                                    Debug.Assert(enumeratorInfo is not null);
+
+                                                                    NamedTypeSymbol intType = rewriter._factory.SpecialType(SpecialType.System_Int32);
+                                                                    if (enumeratorInfo.InlineArraySpanType == WellKnownType.System_Span_T)
+                                                                    {
+                                                                        elementRef = rewriter._factory.ModuleBuilderOpt.EnsureInlineArrayElementRefExists(node.Syntax, intType, rewriter._diagnostics.DiagnosticBag);
+                                                                    }
+                                                                    else
+                                                                    {
+                                                                        Debug.Assert(enumeratorInfo.InlineArraySpanType == WellKnownType.System_ReadOnlySpan_T);
+                                                                        elementRef = rewriter._factory.ModuleBuilderOpt.EnsureInlineArrayElementReadOnlyRefExists(node.Syntax, intType, rewriter._diagnostics.DiagnosticBag);
+                                                                    }
+
+                                                                    TypeSymbol inlineArrayType = boundArrayVar.Type;
+                                                                    elementRef = elementRef.Construct(inlineArrayType, inlineArrayType.TryGetInlineArrayElementField()!.Type);
```

Comment:
>elementRef [](http://example.com/codeflow?start=81&length=10)

Should the `ref TElement` for the 0-th element be calculated in `getPreamble` with `getItem` simply calling `Unsafe.Add()` for the particular index?

Reply:
> Should the code for the 0-th ref TElement be emitted outside the loop in getPreamble, with getItem simply calling Unsafe.Add() to offset for the particular index?

This strategy wouldn't work well for async and iterator scenarios because we will not be able to spill that mutating element ref local. With the current approach, we don't need to be able to spill it. I don't think this possible optimization is worth the added complexity for the mentioned scenarios. We would still need to do the increment and I do not think it makes much difference to increment by one, or by other number. Happy to discuss this off-line in more details. 

> And if so, would the PrivateImplementationDetails methods still be necessary?

We would still need to reinterpret the type of the reference to get the 0-th ref TElement. This is what PrivateImplementationDetails  methods also responsible for.